File: plot/src/main/java/net/fwitz/math/plot/renderer/palette/Interpolate.java
package net.fwitz.math.plot.renderer.palette;

import java.awt.*;

public class Interpolate {
    public static Color interpolate(Color c1, Color c2, double x) {
        return new Color(
                interpolate(c1.getRed(), c2.getRed(), x),
                interpolate(c1.getGreen(), c2.getGreen(), x),
                interpolate(c1.getBlue(), c2.getBlue(), x));
    }

    public static double interpolate(double c1, double c2, double x) {
        return c1 + ((c2 - c1) * x);
    }

    public static int interpolate(int c1, int c2, double x) {
        return c1 + (int) Math.round((c2 - c1) * x);
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/renderer/palette/Interpolate.kt
package net.fwitz.math.plot.renderer.palette

import java.awt.Color
import kotlin.math.roundToInt

object Interpolate {
    fun interpolate(c1: Color, c2: Color, x: Double): Color {
        return Color(
            interpolate(c1.red, c2.red, x),
            interpolate(c1.green, c2.green, x),
            interpolate(c1.blue, c2.blue, x)
        )
    }

    fun interpolate(c1: Double, c2: Double, x: Double): Double {
        return c1 + (c2 - c1) * x
    }

    fun interpolate(c1: Int, c2: Int, x: Double): Int {
        return c1 + ((c2 - c1) * x).roundToInt()
    }
}